python/tskit/stats.py

```diff
@@ -163,3 +167,506 @@ def r2_matrix(self):
+        tree_offset = tree.index
+        num_blocks = min(self.num_blocks, len(tree_idx))
+        tree_blocks = np.floor_divide(num_blocks * tree_idx, len(tree_idx))
+
+        # calculate span weights
+        # TODO: if bootstrapping, does block span need to be tracked
+        # and used to renormalise each replicate?
+        _tree = tree.copy()
+        tree_span = [min(_tree.interval.right, right) - max(_tree.interval.left, left)]
+        while _tree.index < tree_offset + tree_idx[-1]:
+            _tree.next()
+            tree_span.append(
+                min(_tree.interval.right, right) - max(_tree.interval.left, left)
+            )
+        tree_span = np.array(tree_span) / sum(tree_span)
+
+        # storage if using single window, block for entire tree sequence
+        buffer_size = self.buffer_size
+        table_size = buffer_size
+        time = np.zeros(table_size)
+        block = np.zeros(table_size, "int32")
+        weights = np.zeros((table_size, self.num_weights))
+
+        # assemble table of coalescence times in window
+        indices = np.zeros(tree.tree_sequence.num_nodes, "int32") - 1
+        last_block = np.zeros(tree.tree_sequence.num_nodes, "int32") - 1
+        num_record = 0
+        while tree != tskit.NULL:
+            if tree.interval.right > left:
+                current_block = tree_blocks[tree.index - tree_offset]
+                if self.span_normalise:
+                    span_weight = tree_span[tree.index - tree_offset]
+                else:
+                    span_weight = 1.0
+                nodes_in_tree = np.array(
+                    [i for i in tree.nodes() if tree.is_internal(i)]
+                )
+                # TODO this will fail if all nodes are isolated (masked tree)
+                nodes_to_add = nodes_in_tree[
+                    np.where(last_block[nodes_in_tree] != current_block)
+                ]
+                if len(nodes_to_add) > 0:
+                    idx = np.arange(num_record, num_record + len(nodes_to_add))
+                    last_block[nodes_to_add] = current_block
+                    indices[nodes_to_add] = idx
+                    if table_size < num_record + len(nodes_to_add):
+                        table_size += buffer_size
+                        time = np.pad(time, (0, buffer_size))
+                        block = np.pad(block, (0, buffer_size))
+                        weights = np.pad(weights, ((0, buffer_size), (0, 0)))
+                    time[idx] = [tree.time(i) for i in nodes_to_add]
+                    block[idx] = current_block
+                    num_record += len(nodes_to_add)
+                weights[indices[nodes_in_tree], :] += (
+                    span_weight * running_weights[nodes_in_tree, :]
+                )
+
+            if tree.interval.right < right:
+                # if current tree does not cross window boundary, move to next
+                tree.next()
+                self._update_weights_by_edge_diff(
```

Comment:
If we're mutating the input tree anyway here, do we really need to make a full copy above? If the window size isn't too big and we're not moving too many trees along the genome (and the sample size is large) then avoiding a full copy could be a significant perf win.

Reply:
Makes sense! I'll stick in a TODO for now. This "blocking scheme" is going to get revised anyway, because of generalizing the bootstrap machinery.